### PR TITLE
[vcpkg] Use triplet file results in dependency logic expressions

### PR DESCRIPTION
**Pull request by Rastaban**
Previously the logic expression evaluation would do a sub string search of the triplet name to determine if an identifier was 'true' or 'false'.  With this PR it uses the actual values set inside the triplet file.  Most of the changes to vcpkg.exe was refactoring the dependency logic to allow running the triplet file earlier so its results could be used in the qualified dependency resolution.

This PR also fixes the previously undocumented "Supports" CONTROL tag to also use the logic expression evaluation.  Currently the only place the supports tag is used is in the `ci` command.  This will help clean up the results in the CI pipeline - especially when things switch over to using the checked in baseline.  The tag could also be used to provide better error messages on failure, but I will leave that for someone else to do in a later PR.

The unit tests are not complete.

### CODE REVIEW

**Comment by vicroms** [comment]
There's a bug in the dependency graph calculation, transitive dependencies are not being added to the install plan.

Examples:

```cmake
PS D:\src\vcpkg> ./vcpkg install libzippp
The following packages will be built and installed:
  * libzip[core,bzip2,openssl]:x86-windows
    libzippp[core]:x86-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/2: libzip:x86-windows

# Missing bzip and openssl packages in install plan
```

```cmake
PS D:\src\vcpkg> ./vcpkg install opencv
The following packages will be built and installed:
    opencv[dnn,jpeg,webp,opengl,tiff,png,core]:x86-windows
  * opencv4[core,dnn,jpeg,webp,opengl,tiff,png]:x86-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/2: opencv4:x86-windows

# Missing a lot of dependencies
```

**Comment by vicroms** [comment]
It seems to be ignoring `Build-Depends:` completely.
E.g.:

`ports/libzip/CONTROL`
```
Source: libzip
Version: rel-1-5-2
Homepage: https://github.com/nih-at/libzip
Build-Depends: zlib
Default-Features: openssl, bzip2
Description: A library for reading, creating, and modifying zip archives.

Feature: bzip2
Build-Depends: bzip2
Description: Support bzip2-compressed zip archives

Feature: openssl
Build-Depends: openssl
Description: AES (encryption) support using OpenSSL
```

```
PS D:\src\viromer\vcpkg> ./vcpkg install libzip
The following packages will be built and installed:
    libzip[openssl,core,bzip2]:x86-windows
Starting package 1/1: libzip:x86-windows
```

Ignored direct dependency on `zlib` and dependencies from features `bzip2` and `openssl` (which in turn depends on `openssl-windows`).

File: toolsrc/src/vcpkg/build.cpp [hunk 1]
```diff
@@ -62,10 +71,31 @@ namespace vcpkg::Build::Command
...
+        Checks::check_exit(VCPKG_LINE_INFO, feature_dependencies != nullptr);
+        Checks::check_exit(VCPKG_LINE_INFO, package_dependencies != nullptr);
+        Checks::check_exit(VCPKG_LINE_INFO, feature_list != nullptr);
 
-        const Build::BuildPackageConfig build_config{scfl, spec.triplet(), build_package_options, features_as_set};
+        const Build::BuildPackageConfig build_config{scfl,
+                                                     spec.triplet(),
+                                                     build_package_options,
+                                                     var_provider,
+                                                     std::move(*feature_dependencies),
```
Comment (cbezault) [inline toolsrc/src/vcpkg/build.cpp]: These moves can get dropped, BuildPackageConfig takes const references to everything now.

File: toolsrc/src/vcpkg/build.cpp [hunk 2]
```diff
@@ -713,6 +673,14 @@ namespace vcpkg::Build
                                       Hash::Algorithm::Sha1));
```
Comment (cbezault) [inline toolsrc/src/vcpkg/build.cpp]: Above we add the feature_list to the abi file. `config.feature_list` is no longer stable and should be sorted before being added to the abi file.

**Comment by JackBoosY** [comment]
This PR is what we need!

**Comment by cbezault** [comment]
@vicroms, sigh, I probably introduced a bug in one of my last minute changes.

File: toolsrc/src/vcpkg/dependencies.cpp [hunk 3]
```diff
@@ -33,17 +42,131 @@ namespace vcpkg::Dependencies
...
+                    {
+                        if (dep.qualifier.empty())
+                        {
+                            for (const std::string& dep_feature : dep.depend.features)
+                            {
+                                dep_list.emplace_back(
+                                    PackageSpec::from_name_and_triplet(dep.depend.name, m_spec.triplet())
+                                        .value_or_exit(VCPKG_LINE_INFO),
+                                    dep_feature);
+                            }
```
Comment (cbezault) [inline toolsrc/src/vcpkg/dependencies.cpp]: We need to add in the default `FeatureSpec` here. `core` is not in the feature list of `dep.depend.features`.

File: scripts/vcpkg_get_dep_info.cmake [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    # GUID used as a flag - "cut here line"
+    message("c35112b6-d1ba-415b-aa5d-81de856ef8eb")
+    message("VCPKG_TARGET_ARCHITECTURE=${VCPKG_TARGET_ARCHITECTURE}")
+    message("VCPKG_CMAKE_SYSTEM_NAME=${VCPKG_CMAKE_SYSTEM_NAME}")
+    message("VCPKG_CMAKE_SYSTEM_VERSION=${VCPKG_CMAKE_SYSTEM_VERSION}")
+    message("VCPKG_DEP_INFO_OVERRIDE_VARS=${VCPKG_DEP_INFO_OVERRIDE_VARS}")
+    message("CMAKE_HOST_SYSTEM_NAME=${CMAKE_HOST_SYSTEM_NAME}")
+    message("CMAKE_HOST_SYSTEM_PROCESSOR=${CMAKE_HOST_SYSTEM_PROCESSOR}")
+    message("CMAKE_HOST_SYSTEM_VERSION=${CMAKE_HOST_SYSTEM_VERSION}")
+    message("CMAKE_HOST_SYSTEM=${CMAKE_HOST_SYSTEM}")
```
Comment (vicroms) [inline scripts/vcpkg_get_dep_info.cmake]: @cbezault 

Shouldn't `VCPKG_LIBRARY_LINKAGE` and `VCPKG_CRT_LINKAGE` be included here if we want to support the `static` qualifier?

Comment (cbezault) [reply scripts/vcpkg_get_dep_info.cmake]: Yup, oversight on my part.

File: toolsrc/src/vcpkg/dependencies.cpp [hunk 5]
```diff
@@ -33,17 +42,136 @@ namespace vcpkg::Dependencies
...
+                const std::vector<Dependency>* qualified_deps = nullptr;
+                if (feature == "core")
+                {
+                    qualified_deps = &m_scfl.source_control_file->core_paragraph->depends;
+                }
+                else
+                {
+                    qualified_deps =
+                        &m_scfl.source_control_file->find_feature(feature).value_or_exit(VCPKG_LINE_INFO).depends;
+                }
```
Comment (vicroms) [inline toolsrc/src/vcpkg/dependencies.cpp]: This doesn't handle the `*` feature.

**Review by vicroms — CHANGES_REQUESTED** [review]

File: toolsrc/src/vcpkg/dependencies.cpp (hunk 5 above)
Comment (cbezault) [reply toolsrc/src/vcpkg/dependencies.cpp]: It shouldn't have to. The * feature should get expanded before this point. Do you have an example where this misbehaves?

Comment (vicroms) [reply toolsrc/src/vcpkg/dependencies.cpp]: It fails the basic feature tests, any install command should repro: `./vcpkg install cpprestsdk[*]`.

**Comment by vicroms** [comment]
Still failing upgrade plan tests.

**Comment by cbezault** [comment]
I'll take a look when I can.

**Comment by cbezault** [comment]
To summarize our offline conversation. My code doesn't handle the difference in the meaning of the core feature when specified on the command line and when written the the dependency list of a CONTROL file.

File: toolsrc/src/vcpkg/postbuildlint.cpp [hunk 6]
```diff
@@ -98,6 +98,90 @@ namespace vcpkg::PostBuildLint
...
+            "utime.h",     "utmp.h",        "utmpx.h",        "values.h",       "wait.h",      "wchar.h",
+            "wctype.h",    "wordexp.h",
+        };
+        // These files are general names that have shown to be problematic in the past
+        static constexpr StringLiteral restricted_general_filenames[] = {
+            "json.h",
+            "parser.h",
+            "lexer.h",
+        };
+        static constexpr Span<const StringLiteral> restricted_lists[] = {
```
Comment (vicroms) [inline toolsrc/src/vcpkg/postbuildlint.cpp]: Shouldn't we add C++ Standard Library header names as well? At least one library (`sfsexp`) is known to install `cstring.h` directly into `include/`.

**Comment by dan-shaw** [comment]
This looks like it has passed the CI. Before merging this, can we add more documentation as to what was changed (such as adding a new build policy)?

**Comment by ras0219-msft** [comment]
Also, @winsoft666, as part of this PR I discovered that `teemo` installing `include/slice.h` caused x264 to fail to build (and it's easy to see that `#include "slice.h"` is likely to conflict with other libraries).

I resolved this by moving the includes into `include/teemo/` -- CMake projects should be unaffected and can continue using the same include paths, but MSBuild projects will now need to use `#include <teemo/teemo.h>`. I'd strongly recommend considering making this change upstream and having all users always include `teemo/teemo.h`.

File: docs/maintainers/control-files.md [hunk 7]
```diff
@@ -96,38 +96,55 @@ Vcpkg does not distinguish between build-only dependencies and runtime dependenc
...
 This field is optional.
 
+##### Example:
 ```no-highlight
 Default-Features: dynamodb, s3, kinesis
 ```
 
 #### Supports
-Expression that evalutates to true when the port is supported for a given triplet.  This field is currently only used in the CI testing to skip ports.
+Expression that evalutates to true when the port is expected to build successfully for a triplet.
```
Comment (cbezault) [inline docs/maintainers/control-files.md]: Evaluates

**Review by vicroms — APPROVED** [review]
Let's Get This Merged

**Comment by cbezault** [comment]
bless

**Comment by JackBoosY** [comment]
With this PR, I think we can further improve the CI test to add feature tests.

**Pull request merged**